Add TypeScript type tests for check, random, email, fetch and tracker packages

gauravarora opened the pull request:
## Summary
- Addresses issue #13676 by adding TypeScript type tests to verify type definitions work correctly
- Establishes a pattern for adding type tests to other packages with `.d.ts` files
- Added comprehensive type tests for 5 packages: `check`, `random`, `tracker`, `fetch`, and `email`

## Changes
- **check**: Added type tests for Match patterns (Maybe, Optional, OneOf, ObjectIncluding, Where, Integer, Any) and check() assertions
- **random**: Added type tests for all Random methods (id, secret, fraction, hexString, choice)
- **tracker**: Added type tests for Tracker.autorun, Computation interface, Dependency class, and reactive utilities
- **fetch**: Added type tests for fetch(), Headers, Request, and Response APIs
- **email**: Added type tests for Email.sendAsync(), hookSend(), customTransport(), and MailComposer

## Pattern for other packages
Each package now follows this pattern:
1. Add `'typescript'` to `Package.onTest()` dependencies
2. Create a `*_test.ts` file with type tests using `api.mainModule()`
3. Import APIs and verify type signatures, return types, and type narrowing

Fixes #13676

gauravarora committed:
Add TypeScript type tests for check, random, and tracker packages

Addresses issue #13676 by adding TypeScript type tests to verify type definitions work correctly. This establishes a pattern for adding type tests to other packages with .d.ts files.

Changes:
- check: Added comprehensive type tests for Match patterns and check() assertions
- random: Added type tests for all Random methods (id, secret, fraction, hexString, choice)
- tracker: Added type tests for Tracker.autorun, Computation, Dependency, and reactive utilities

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/meteor/meteor?pullRequest=13951) <br/>All committers have signed the CLA.

gauravarora committed:
Add TypeScript type tests for fetch and email packages

- fetch: Added type tests for fetch(), Headers, Request, and Response APIs
- email: Added type tests for Email.sendAsync(), hookSend(), customTransport(), and MailComposer

italojs committed:
Merge branch 'devel' into add-typescript-type-tests

gauravarora committed:
Merge branch 'devel' into add-typescript-type-tests

italojs commented on packages/check/match_test.ts:
```diff
@@ -0,0 +1,99 @@
+import { Tinytest } from "meteor/tinytest";
+import { check, Match } from "meteor/check";
+
+Tinytest.add("Check - TypeScript types - basic check function", (test) => {
```
would be better to have a test for "general types" so test string, number, obj...

italojs commented on packages/check/match_test.ts:
```diff
@@ -0,0 +1,99 @@
+  test.equal(upper, "HELLO");
+});
+
+Tinytest.add("Check - TypeScript types - Match.test with primitives", (test) => {
```
[Check](https://docs.meteor.com/api/check) lib isnt a typescript feature, all this file have `TypeScript types` in the test descriptions, bu none of them is testing typescript related things, `Number and String` is js native

italojs commented on packages/check/package.js:
```diff
@@ -16,7 +16,8 @@ Package.onUse(api => {
+  api.use(['check', 'tinytest', 'ejson', 'ecmascript', 'typescript'], ['client', 'server']);
 
   api.addFiles('match_test.js', ['client', 'server']);
+  api.mainModule('match_test.ts');
```
the file is .ts but isnt using ts statement 🤷🏻‍♂️
vote to remove it

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+
+Tinytest.add("Email - TypeScript types - sendAsync options", (test) => {
+  // Type check that sendAsync accepts proper options
+  const options = {
```
is missing the type `EmailOptions`

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+});
+
+Tinytest.add("Email - TypeScript types - hookSend", (test) => {
+  Email.hookSend((options) => {
```
missing the type EmailOptions into options `Email.hookSend((options: EmailOptions) => {`

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+    return true;
+  });
+
+  test.equal(true, true);
```
this line doest make sense, you can use test. use `test.ok(...)` istead

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+    forceEmbeddedImages: false
+  });
+
+  test.equal(typeof composer, "object");
```
it should be typeof composer, "MailComposer", right?

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+  });
+
+  composer.addHeader("X-Custom-Header", "value");
+  test.equal(typeof composer.addHeader, "function");
```
why not merge this test with the above one?

italojs commented on packages/fetch/tests/main.ts:
```diff
@@ -0,0 +1,95 @@
+import { Tinytest } from "meteor/tinytest";
+import { fetch, Headers, Request, Response } from "meteor/fetch";
+
+Tinytest.addAsync("Fetch - TypeScript types - fetch function", async (test, done) => {
```
here we are testing node's fetch, not a method stuff, remove all tests like this, please
featch already have your own tests

italojs commented on packages/tracker/tracker_tests.ts:
```diff
@@ -0,0 +1,185 @@
+});
+
+Tinytest.add("Tracker - TypeScript types - Computation properties", (test) => {
+  const computation = Tracker.autorun((c) => {
```
missing the type `...: Tracker.Computation =...`

italojs commented on packages/tracker/tracker_tests.ts:
```diff
@@ -0,0 +1,185 @@
+    const stopped: boolean = c.stopped;
+    const promise: Promise<unknown> = c.firstRunPromise;
+
+    test.equal(typeof firstRun, "boolean");
```
it's typescript, [Boolean](https://www.tutorialsteacher.com/typescript/typescript-boolean) doent work here?

italojs commented on packages/email/email_tests.ts:
```diff
@@ -0,0 +1,106 @@
+
+  // This ensures the type signature is correct
+  const sendPromise: Promise<void> = Email.sendAsync(options);
+  test.equal(typeof sendPromise.then, "function");
```
it's typescript, `instanceof Promise` doesnt work here?

italojs commented on packages/tracker/tracker_tests.ts:
```diff
@@ -0,0 +1,185 @@
+  test.equal(computation.invalidated, true);
+
+  computation.onInvalidate((c: Tracker.Computation) => {
+    test.equal(typeof c, "object");
```
should be `typeof c, Tracker.Computation`

italojs commented:
@gauravarora, any update?

Grubba27 commented:
We can reopen this if you want to continue the work!
